### PR TITLE
Section 4 - SPS Project Submission (Part 1 and 2)

### PR DESCRIPTION
**Pull request by jiang-lim**
Please fill out the survey before submitting the pull request. Thanks!

🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

How many hours did you spend on this assignment?

~1

Please fill in one error and/or error message you received while working on this assignment.

N/A

What part of the assignment did you spend the most time on?

Thinking about the approach.

Comfort Level (1-5):

4

Completeness Level (1-5):

5

What did you think of this deliverable?

It was fun! Enjoyed thinking about the approach for the reversed version.

Is there anything in this code that you feel pleased about?

The reversed version.

### CODE REVIEW

**Comment by jiang-lim** [comment]
**[Survey for Part 2]**

Please fill out the survey before submitting the pull request. Thanks!

🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

How many hours did you spend on this assignment?

~4 for all the Base + More Comfortable exercises.

Please fill in one error and/or error message you received while working on this assignment.

One of the values was returned as "undefined", so I had to re-check my function logic.

What part of the assignment did you spend the most time on?

Thinking about the approach for the Korean mode. Initially over-complicated the logic but eventually arrived at a straightforward and simple approach!

Also ended up wasting quite a lot of time re-factoring the entire logic upon each new requirement. I should have just reviewed all requirements (in Base + More Comfortable) before starting work.

Comfort Level (1-5):

3.5

Completeness Level (1-5):

5

What did you think of this deliverable?

It was challenging but really fun! Enjoyed thinking about the approach for the Korean version.

Is there anything in this code that you feel pleased about?

Segregating the logic for each different game from the main function.

**Comment by jiang-lim** [comment]
The current logic in the `main` function is quite messy. Would appreciate any feedback on improving that!

File: script.js [hunk 1]
```diff
@@ -1,4 +1,190 @@
...
+    userChoice = generateRandomChoice();
+  } else {
+    userChoice = userMove;
+  }
+  if (userChoice == progChoice) {
+    gameResult = DRAW;
+  } else if (
+    (userChoice == SCISSORS && progChoice == STONE) ||
+    (userChoice == PAPER && progChoice == SCISSORS) ||
+    (userChoice == STONE && progChoice == PAPER)
```
Comment (zephaniahong) [inline script.js:71]: This logic seems to be repeated several times in your code. This is a tell tale sign that you could consider abstracting it into a separate function.

File: script.js [hunk 2]
```diff
@@ -1,4 +1,190 @@
...
+      myOutputValue = `You have selected ${COM_VS_COM} mode! Click on "Submit" to play.`;
+    } else if (input == KOREAN) {
+      gameMode = KOREAN;
+      myOutputValue = `You have selected ${KOREAN} mode! Type "${SCISSORS}", "${PAPER}" or "${STONE}" to play.`;
+    } else {
+      myOutputValue = `Please type "${NORMAL}", "${REVERSE}", "${COM_VS_COM}" or "${KOREAN}" to select the game mode.`;
+    }
+  } else if (gameMode == NORMAL) {
+    if (input != SCISSORS && input != PAPER && input != STONE) {
+      myOutputValue = `Please type "${SCISSORS}", "${PAPER}" or "${STONE}" to play normal mode.`;
```
Comment (zephaniahong) [inline script.js:170]: Consider abstracting the generating of the outputvalue message to a separate function.

File: script.js [hunk 3]
```diff
@@ -1,4 +1,190 @@
...
+    } else {
+      myOutputValue = interpretResult(reverseSPS(input));
+    }
+  } else if (gameMode == COM_VS_COM) {
+    myOutputValue = interpretResult(basicSPS());
+  } else if (gameMode == KOREAN) {
+    if (input != SCISSORS && input != PAPER && input != STONE) {
+      myOutputValue = `Please type "${SCISSORS}", "${PAPER}" or "${STONE}" to play korean mode.`;
+    } else {
+      myOutputValue = interpretResult(koreanSPS(input));
```
Comment (zephaniahong) [inline script.js:186]: This is quite an advanced way of using functions but great job on that!(:

File: script.js [hunk 4]
```diff
@@ -1,4 +1,190 @@
...
+    latestWinner = currentResult;
+    gameResult = PENDING_DRAW;
+  }
+  console.log(`current result ${currentResult}`);
+  console.log(`game result ${gameResult}`);
+  console.log(`user win ${USER_WIN}`);
+  return gameResult;
+};
+
+var interpretResult = function (gameResult) {
```
Comment (zephaniahong) [inline script.js:112]: Good function name.

**Review by zephaniahong — COMMENTED** [review]
Overall a very good job. I do not see much issues with your code. You have a very strong understanding of functions. One small tip would be to spend some time after completing the project to see where there are repeated chunks of code and consider abstracting that out into a separate function. Other than that, a really solid job!(👍

**Pull request closed without merging**